lib/elixir/lib/keyword.ex

```diff
@@ -80,16 +80,19 @@ defmodule Keyword do
 
   ## Examples
 
-      iex> Keyword.new([:a, :b], fn (x) -> {x, x} end) |> Enum.sort
-      [a: :a, b: :b]
+      iex> Keyword.new([:a, :b], fn (x) -> {x, x} end)
+      [b: :b, a: :a]
 
   """
-  @spec new(Enum.t, ({key, value} -> {key, value})) :: t
-  def new(pairs, transform) do
-    Enum.reduce pairs, [], fn i, keywords ->
-      {k, v} = transform.(i)
-      put(keywords, k, v)
+  @spec new(Enum.t, (term -> {key, value})) :: t
+  def new(pairs, transform) when is_function(transform, 1) do
+    fun = fn el, acc ->
+      {k, v} = transform.(el)
+      put_new(acc, k, v)
     end
+
+    keywords = :lists.foldl(fun, [], Enum.reverse(pairs))
```

Comment:
Right, there is no need for tail recursion since keywords shouldn't be enormously big.